### PR TITLE
Migration source and target implementations

### PR DESCRIPTION
**Pull request by sergenyalcin**
### Description of your changes

Depends on https://github.com/upbound/upjet/pull/119
Related https://github.com/upbound/team-extensions/issues/29

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Tested as part of upbound/team-control-planes#119 

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/migration/filesystem.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	if o.Metadata.Parents != "" {
+		f, err := ft.afero.OpenFile(o.Metadata.Path, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+		if err != nil {
+			return err
+		}
+
+		defer f.Close() //nolint:errcheck
+
+		if _, err = f.WriteString(fmt.Sprintf("\n---\n\n%s", string(b))); err != nil {
```
Comment (ulucinar) [inline pkg/migration/filesystem.go:134]: So the idea is, if a migration converter produces multiple managed resources, then all will be appended to the same file, right?

File: pkg/migration/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+// KubernetesSource is a source implementation to read resources from Kubernetes
+// cluster.
+type KubernetesSource struct {
+	index         int
+	items         []UnstructuredWithMetadata
+	dynamicClient dynamic.Interface
+}
+
+// NewKubernetesSource returns a KubernetesSource
+func NewKubernetesSource(dynamicClient dynamic.Interface, gvks []schema.GroupVersionKind) (*KubernetesSource, error) {
```
Comment (ulucinar) [inline pkg/migration/kubernetes.go:23]: Documentation on the parameters (especially the `gvks` parameter) would be good.

File: pkg/migration/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	return ks, nil
+}
+
+// HasNext checks the next item
+func (ks *KubernetesSource) HasNext() (bool, error) {
+	if ks.index < len(ks.items) {
+		return true, nil
+	}
+	return false, nil
```
Comment (ulucinar) [inline pkg/migration/kubernetes.go:56]: nit:
```suggestion
	return ks.index < len(ks.items), nil
```

File: pkg/migration/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	if hasNext, _ := ks.HasNext(); hasNext {
+		item := ks.items[ks.index]
+		ks.index++
+		return item, nil
+	}
+	return UnstructuredWithMetadata{}, errors.New("failed to get next element")
+}
+
+// InitializeDynamicClient returns a dynamic client
+func InitializeDynamicClient(kubeconfigPath string) (dynamic.Interface, error) {
```
Comment (ulucinar) [inline pkg/migration/kubernetes.go:70]: We consider this as a utility to construct dynamic clients, is this correct?

File: pkg/migration/filesystem.go (hunk 1 above)
Comment (sergenyalcin) [reply pkg/migration/filesystem.go:132]: Yes, right! This block adds `---` between manifests.

**Review by ulucinar — APPROVED** [review]
Thanks @sergenyalcin for the implementations. I think we are ready to merge them and do some further iterations as we gain more experience on the migration use cases and corner cases. 

Because you are busy with support rotation, I can take care of rebasing the branch and merging it as is.

**Comment by ulucinar** [comment]
Hi @sergenyalcin,
I have rebased your PR. I only did the following changes in your feature branch:
- Moved test resources `awsvpc.yaml` and `resourcegroup.yaml` from `testdata` to `testdata/source`.
- Added the following error check in the test file `pkg/migration/filesystem_test.go`:
```go
			fs, err := NewFileSystemSource("testdata/source", FsWithFileSystem(files))
			if err != nil {
				t.Fatalf("Failed to initialize a new FileSystemSource: %v", err)
			}
```
- Updated references to `testdata` with `testdata/source`.

**Pull request merged**